### DISCUSSION TITLE
What is the most insightful large scale spike analysis that could be run on the 1million cells of multi_area_model?

### DISCUSSION THREAD

**Discussion by russelljjarvis** [genn-team/genn#526] [Ideas]
https://github.com/BrainsOnBoard/procedural_paper

Use Julia Packages: https://github.com/lindermanlab/PPSeq.jl, https://github.com/dpshorten/CoTETE.jl, OnlineStats.jl
https://github.com/JuliaNeuroscience/SpikeSynchrony.jl

To analyse the 1million cells of the multi_area_model spiking data at large scale.

This is a technical area involving several points of curiosity.

* Does this choice of analysis algorithm scale okay?

* Can algorithms be re-written to work in an online fashion? Ie, not the whole data-set is stored in memory, but chunks are operated on in a rotating buffer type manner. 

* Can GPU acceleration be used (Julia provides nice GPU libraries for CuArrays and for writing custom kernels).

If not does operate on chunks of data and provides valuable approximations.

https://github.com/russelljjarvis/PPSeq.jl/discussions/1

**Comment by russelljjarvis** [comment]
Hi @Tijl and @aMarcireau,

This is a spike train is the output of 1million cells from a scaled down version of the [Multi-scale spiking network model of macaque visual cortex](https://github.com/INM-6/multi-area-model)  (full scale 100 million).

![image](https://user-images.githubusercontent.com/7786645/171562816-51863e9c-23cf-4a34-b33d-172e1183ec45.png)

Goal: do Representational Dissimilarity Analysis between the simulation outputs and fMRI voxel activations in V1 in a virtual experiment where both networks are stimulated by the same stimulus.

Problem 1: Kreutz spike distance is the most natural measure of spike train similarity, computing the Kreutz spike distance between each possible pairing of cells in a 1million neuron simulation, leads to filling a matrix with $ 1million^{2} $ elements, a calculation that takes 4 days to compute, and that won't be easily reproducible.

Problem 2: Although PCA is a good dimensionality reduction technique. Finding a projection space that maximises variance in the data, is going to upward bias the found dissimilarity, because the spike trains returned by a PCA fitting routine are going to be the ones that represent high variance, and each one is going to come from an orthogonal projection vector, so each spike train is going to look very dissimilar to the other.

Alternative idea, perform n=10 random subsamples of ncell=250 into the 1million neuron spike train, compute Kreutz spike distance matrix on the $ 250^{2} $ element matrix 10 times. Use the 10 samples to do a hypothesis test about the dissimilarity.


### Other issues:
* There is a handful of off the shelf Python packages that do dimensionality reduction on spike trains, but rather than outputting a smaller number of spike trains, these algorithms tend to output smooth lines. Ie I can do a dimensionality reduction that converts spikes to [smooth lines](https://github.com/catniplab/vlgp/blob/master/notebook/tutorial.ipynb)
* Furthermore, there is less reason to believe these algorithms written in numpy will scale better  or worse than algorithms written in Julia.
But what is really wanted is large number of spike trains -> smaller number of representative spike trains.


### Interesting but Un-related:

I can divide it into chunks of 750 cells and sort the raster spike plots to increase cluster sizes of coincident events

This chunk of 750 cells is 
![image](https://user-images.githubusercontent.com/7786645/171563018-67e8837e-d5c0-4f5d-819c-d9c6947ae2e1.png)

transformed to this chunk:

![image](https://user-images.githubusercontent.com/7786645/171562992-62738184-d175-4f1f-9d44-7b98b45abde3.png)

**Comment by russelljjarvis** [comment]
@Tijl,
It just occurred to me, this a model that contains many more cell populations than V1. I can use domain knowledge to guide subsampling of cells, firstly I should only sample from v1, secondly I should only sample from a particular layer of V1, layer IV might be the appropriate readout layer.

**Reply by Tijl** [reply]
your problem#1 is not exactly right. We just need to calculate the distance between two of such raster plots (i.e. in response to two different images). So the distance would be cell by cell, but not all pairs. Does that make things easier?

**Reply by russelljjarvis** [reply]
Definitely! That will be a significant reduction of computations ie $ 1 \times 10^{6} $ as opposed to $ 1 \times 10^{12} $

**Reply by russelljjarvis** [reply]
It turns out that the inputs to this large visual cortex model is just synaptic noise (as synaptic drive).
https://github.com/orgs/genn-team/discussions/529#discussioncomment-2896430

I think the simulation is not a simulation of active visual perception, but something else (self sustaining network activity, day dreaming etc).
